File: library/src/test/java/io/ashdavies/rx/rxtasks/SingleTaskOnSubscribeTest.java
package io.ashdavies.rx.rxtasks;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;

@RunWith(MockitoJUnitRunner.class)
public class SingleTaskOnSubscribeTest {

  private SingleOnSubscribe<String> onSubscribe;

  @Mock Task<String> task;
  @Mock SingleEmitter<String> emitter;

  @Mock TaskListenerFactory<String, SingleEmitter<String>> factory;
  @Mock OnSuccessListener<String> onSuccessListener;
  @Mock OnFailureListener onFailureListener;

  @Before
  public void setUp() throws Exception {
    onSubscribe = new SingleTaskOnSubscribe<>(task, factory);
  }

  @Test
  public void shouldSubscribeWithOnSuccessListener() throws Exception {
    given(factory.createOnSuccessListener(emitter)).willReturn(onSuccessListener);

    onSubscribe.subscribe(emitter);

    then(factory).should().createOnSuccessListener(emitter);
    then(task).should().addOnSuccessListener(onSuccessListener);
  }

  @Test
  public void shouldSubscribeWithOnFailureListener() throws Exception {
    given(factory.createOnFailureListener(emitter)).willReturn(onFailureListener);

    onSubscribe.subscribe(emitter);

    then(factory).should().createOnSuccessListener(emitter);
    then(task).should().addOnFailureListener(onFailureListener);
  }
}

File: library/src/test/kotlin/io/ashdavies/rx/rxtasks/SingleTaskOnSubscribeTest.kt
package io.ashdavies.rx.rxtasks

import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.android.gms.tasks.Task
import com.nhaarman.mockito_kotlin.argumentCaptor
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.given
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.then
import io.reactivex.SingleEmitter
import org.junit.Test
import java.util.concurrent.Executor

internal class SingleTaskOnSubscribeTest {

  private val task = mock<Task<Boolean>>()
  private val emitter = mock<SingleEmitter<Boolean>>()
  private val executor = mock<Executor>()

  private val single = SingleTaskOnSubscribe(task, executor)

  @Test
  fun `should add on success listener`() {
    val captor = argumentCaptor<OnSuccessListener<Boolean>>()
    given(task.addOnSuccessListener(eq(executor), captor.capture())).willReturn(task)

    single.subscribe(emitter)
    captor.lastValue.onSuccess(true)

    then(emitter).should().onSuccess(true)
  }

  @Test
  fun `should add on failure listener`() {
    val captor = argumentCaptor<OnFailureListener>()
    given(task.addOnFailureListener(eq(executor), captor.capture())).willReturn(task)

    val exception = RuntimeException()
    single.subscribe(emitter)

    captor.lastValue.onFailure(exception)
    then(emitter).should().onError(exception)
  }
}
